quarterback-packages/api/src/routes/doc/doc.ctrl.ts

```diff
@@ -101,3 +103,150 @@ export const deleteDocument = async (
+}
+
+export const receiveSteps = async (
+  req: AuthRequest<any>,
+  res: AuthResponse,
+  next: NextFunction
+) => {
+  try {
+    const { documentId } = req.params
+    const newSteps = req.body.steps
+    const clientID = req.body.clientID // clientID is not the same as userID, it's an ID of the client exactly (app instance running in the browser)
+    const clientVersion = req.body.version // client the version on top of which it updates, in other words it says last version known to it
+
+    const docSearch = await docService.findDocument(documentId)
+    if ('data' in docSearch) {
+      // @ts-ignore
+      const { version } = docSearch.data
+
+      // this is needed, commented for testing only
+      if (version != clientVersion) {
+        /*
+        if client is based on an older version we deny his request. At some point in time client should receive update and will then
+        resend his steps after rebasing them. Client will persist its steps.
+        */
+        return
+      }
+
+      const cachedItem = getDocHistory(documentId)
+
+      let pmDoc = cachedItem.doc || schema.nodeFromJSON(docSearch.data.doc)
+
+      newSteps.forEach((jsonStep) => {
+        const step = Step.fromJSON(schema, jsonStep)
+        pmDoc = step.apply(pmDoc).doc || pmDoc
+        cachedItem.steps.push(step)
+        cachedItem.clientIDs.push(clientID)
+      })
+      cachedItem.version = clientVersion
+      cachedItem.doc = pmDoc
+
+      // docService.updateDocument(documentId, {doc: pmDoc.toJSON()})
+      res.sendStatus(200)
+
+      // Signal server side listener
+      sendStepsToAll({ steps: newSteps, clientIDs: cachedItem.clientIDs, version: clientVersion })
+    } else {
+      next(new CustomError('Unable to save steps', 500))
+    }
+  } catch (err) {
+    next(err)
+  }
+}
+
+function sendStepsToAll(data: unknown) {
+  global.clients.forEach((client) => client.res.write(`data: ${JSON.stringify(data)}\n\n`))
+}
+
+export const stepsEventHandler = (
+  req: AuthRequest,
+  res: AuthResponse<string>,
+  next: NextFunction
+) => {
+  const headers = {
+    'Content-Type': 'text/event-stream',
+    Connection: 'keep-alive',
+    'Cache-Control': 'no-cache',
+  }
+
+  // initial response
+  const { documentId } = req.params
+  const historyData = getDocHistory(documentId)
+  const data = `data: ${JSON.stringify(historyData)}\n\n`
+  res.writeHead(200, headers)
+  res.write(data)
+
+  const clientId = Date.now()
+
+  const newClient = {
+    id: clientId,
+    res,
+  }
+  global.clients.push(newClient)
+
+  req.on('close', () => {
+    console.log(`${clientId} Connection closed`)
+    const index = global.clients.findIndex((client) => client.id == clientId)
+    global.clients.splice(index)
+  })
+}
+
+type StepsSince = {
+  steps: unknown[] // json representaion of Step[]
+  clientIDs: unknown[] // string or integer
+  version: number
+}
+
+export const getVersion = (req: AuthRequest, res: AuthResponse<StepsSince>, next: NextFunction) => {
```

Comment:
@mbetamony , when FE gets notified via SSE about new steps on the document, it will either use the steps from the notification or will get steps since the version it currently has using this controller.